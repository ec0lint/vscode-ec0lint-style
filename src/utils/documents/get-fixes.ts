import { createTextEdits } from './create-text-edits';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
import type { Ec0lintRunner, RunnerOptions } from '../ec0lint-style';

/**
 * Runs Ec0lint and returns fix text edits for the given document.
 * @param runner The Ec0lint runner.
 * @param document The document to get fixes for.
 * @param linterOptions Linter options to use.
 * @param runnerOptions The runner options.
 */
export async function getFixes(
	runner: Ec0lintRunner,
	document: TextDocument,
	linterOptions: Ec0lintStyle.LinterOptions = {},
	runnerOptions: RunnerOptions = {},
): Promise<TextEdit[]> {
	const result = await runner.lintDocument(
		document,
		{ ...linterOptions, fix: true },
		runnerOptions,
	);

	return typeof result.output === 'string' ? createTextEdits(document, result.output) : [];
}
