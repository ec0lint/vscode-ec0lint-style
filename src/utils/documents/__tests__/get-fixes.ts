import { TextDocument } from 'vscode-languageserver-textdocument';
import type Ec0lintStyle from 'ec0lint-style';
import type { StylelintRunner } from '../../stylelint';
import type { RunnerOptions } from '../../stylelint';

import { getFixes } from '../get-fixes';

const createMockRunner = (output?: string) =>
	({
		lintDocument: jest.fn(async () => ({ diagnostics: [], output })),
	} as unknown as StylelintRunner);

const createDocument = (code: string) =>
	TextDocument.create('file:///path/test.css', 'css', 1, code);

describe('getFixes', () => {
	test('should call lintDocument with given options and fix set to true', async () => {
		const document = createDocument('a { color: red; }');
		const runner = createMockRunner('a { color: red; }');
		const linterOptions: Ec0lintStyle.LinterOptions = {
			config: {
				customSyntax: 'postcss-scss',
			},
			fix: false,
			reportDescriptionlessDisables: false,
			reportNeedlessDisables: true,
			reportInvalidScopeDisables: false,
		};

		const runnerOptions: RunnerOptions = {
			packageManager: 'pnpm',
			validate: ['css', 'less', 'sass', 'scss'],
		};

		await getFixes(runner, document, linterOptions, runnerOptions);

		expect(runner.lintDocument).toHaveBeenCalledWith(
			document,
			expect.objectContaining({
				...linterOptions,
				fix: true,
			}),
			runnerOptions,
		);
	});

	test('should return no edits if Stylelint returned no output', async () => {
		const document = createDocument(`
			a {
				color: red;
			}

			div#foo {
			}

			.foo {
				overflow: hidden;
				background: #ccc url(foo.png) no-repeat;
			}
		`);
		const runner = createMockRunner();
		const fixes = await getFixes(runner, document);

		expect(fixes).toEqual([]);
	});

	test('should return no edits if Stylelint made no changes', async () => {
		const document = createDocument(`
			a {
				color: red;
			}

			div#foo {
			}

			.foo {
				overflow: hidden;
				background: #ccc url(foo.png) no-repeat;
			}
		`);
		const runner = createMockRunner(`
			a {
				color: red;
			}

			div#foo {
			}

			.foo {
				overflow: hidden;
				background: #ccc url(foo.png) no-repeat;
			}
		`);
		const fixes = await getFixes(runner, document);

		expect(fixes).toEqual([]);
	});

	test('should return edits if Stylelint made changes', async () => {
		const document = createDocument(`
			a {
				color: red;
			}

			div#bar {
			}

			.foo {
				overflow: hidden;
				background: #ccc url(foo.png) no-repeat;
			}
		`);
		const runner = createMockRunner(`
			a {
			  color: red;
			}

			.foo {
			  overflow: hidden;
			  background: #ccc url(foo.png) no-repeat;
			}
		`);
		const fixes = await getFixes(runner, document);

		expect(fixes).toMatchSnapshot();
	});
});
