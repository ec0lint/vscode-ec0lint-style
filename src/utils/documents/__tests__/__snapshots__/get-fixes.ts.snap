// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFixes should return edits if Ec0lint made changes 1`] = `
Array [
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 2,
      },
      "start": Object {
        "character": 3,
        "line": 2,
      },
    },
  },
  Object {
    "newText": "  ",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 2,
      },
      "start": Object {
        "character": 4,
        "line": 2,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 3,
        "line": 8,
      },
      "start": Object {
        "character": 3,
        "line": 5,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 9,
      },
      "start": Object {
        "character": 3,
        "line": 9,
      },
    },
  },
  Object {
    "newText": "  ",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 9,
      },
      "start": Object {
        "character": 4,
        "line": 9,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 10,
      },
      "start": Object {
        "character": 3,
        "line": 10,
      },
    },
  },
  Object {
    "newText": "  ",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 10,
      },
      "start": Object {
        "character": 4,
        "line": 10,
      },
    },
  },
]
`;

exports[`getFixes should return edits if Stylelint made changes 1`] = `
Array [
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 2,
      },
      "start": Object {
        "character": 3,
        "line": 2,
      },
    },
  },
  Object {
    "newText": "  ",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 2,
      },
      "start": Object {
        "character": 4,
        "line": 2,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 3,
        "line": 8,
      },
      "start": Object {
        "character": 3,
        "line": 5,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 9,
      },
      "start": Object {
        "character": 3,
        "line": 9,
      },
    },
  },
  Object {
    "newText": "  ",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 9,
      },
      "start": Object {
        "character": 4,
        "line": 9,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 10,
      },
      "start": Object {
        "character": 3,
        "line": 10,
      },
    },
  },
  Object {
    "newText": "  ",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 10,
      },
      "start": Object {
        "character": 4,
        "line": 10,
      },
    },
  },
]
`;
